examples/gossipsub-chat.rs

```diff
@@ -18,141 +18,149 @@
-            .validation_mode(ValidationMode::Strict) // This sets the kind of message validation. The default is Strict (enforce message signing)
-            .message_id_fn(message_id_fn) // content-address messages. No two messages of the
-            // same content will be propagated.
-            .build()
-            .expect("Valid config");
-        // build a gossipsub network behaviour
-        let mut gossipsub: gossipsub::Gossipsub =
-            gossipsub::Gossipsub::new(MessageAuthenticity::Signed(local_key), gossipsub_config)
-                .expect("Correct configuration");
-
-        // subscribes to our topic
-        gossipsub.subscribe(&topic).unwrap();
-
-        // add an explicit peer if one was provided
-        if let Some(explicit) = std::env::args().nth(2) {
-            match explicit.parse() {
-                Ok(id) => gossipsub.add_explicit_peer(&id),
-                Err(err) => println!("Failed to parse explicit peer id: {:?}", err),
-            }
-        }
-
-        // build the swarm
-        libp2p::Swarm::new(transport, gossipsub, local_peer_id)
+        let mdns = Mdns::new(MdnsConfig::default())?;
+        let behaviour = MyBehaviour { gossipsub, mdns };
+        Swarm::new(transport, behaviour, local_peer_id)
     };
 
-    // Listen on all interfaces and whatever port the OS assigns
-    swarm
-        .listen_on("/ip4/0.0.0.0/tcp/0".parse().unwrap())
-        .unwrap();
-
-    // Reach out to another node if specified
-    if let Some(to_dial) = std::env::args().nth(1) {
-        let address: Multiaddr = to_dial.parse().expect("User to provide valid address.");
-        match swarm.dial(address.clone()) {
-            Ok(_) => println!("Dialed {:?}", address),
-            Err(e) => println!("Dial {:?} failed: {:?}", address, e),
-        };
-    }
-
     // Read full lines from stdin
     let mut stdin = io::BufReader::new(io::stdin()).lines().fuse();
 
+    // Listen on all interfaces and whatever port the OS assigns
+    swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?;
+
+    println!("Enter messages via STDIN and they will be sent to connected peers using Gossipsub");
+
     // Kick it off
     loop {
         select! {
             line = stdin.select_next_some() => {
                 if let Err(e) = swarm
-                    .behaviour_mut()
-                    .publish(topic.clone(), line.expect("Stdin not to close").as_bytes())
-                {
+                    .behaviour_mut().gossipsub
+                    .publish(topic.clone(), line.expect("Stdin not to close").as_bytes()) {
                     println!("Publish error: {:?}", e);
                 }
             },
             event = swarm.select_next_some() => match event {
-                SwarmEvent::Behaviour(GossipsubEvent::Message {
+                SwarmEvent::Behaviour(MyBehaviourEvent::Mdns(MdnsEvent::Discovered(list))) => {
```

Comment:
Not required but might want to use check for `MdnsEvent::Expired` to remove the explicit peer as well.

Reply:
Nice. Does that event occur after DNS TTL or some other cause?

While using this code in my own project, I wanted to look at other events I should be processing to make it more complete and robust.

Is there any peer event when the other peer is killed with Control-C?

If you agree, I'd like to look at that a bit more and follow-up with a second PR that makes the example more complete and robust?

Reply:
OK, I see from https://docs.rs/libp2p/0.18.1/libp2p/mdns/enum.MdnsEvent.html it is on TTL and non-renewal.

Reply:
I ended up just adding that to the PR as it was straightforward. Thanks for the improvement suggestion.
I also rebased on master.
OK to merge now? Additional reviews required to merge?